src/backend/distributed/commands/multi_copy.c

```diff
@@ -0,0 +1,1284 @@
+		PGresult *result = NULL;
+
+		if (transactionConnection->transactionState != TRANSACTION_STATE_COPY_STARTED)
+		{
+			/* a failure occurred after having previously called EndRemoteCopy */
+			continue;
+		}
+
+		/* end the COPY input */
+		copyEndResult = PQputCopyEnd(connection, NULL);
+		transactionConnection->transactionState = TRANSACTION_STATE_OPEN;
+
+		if (copyEndResult != 1)
+		{
+			if (stopOnFailure)
+			{
+				char *nodeName = ConnectionGetOptionValue(connection, "host");
+				char *nodePort = ConnectionGetOptionValue(connection, "port");
+
+				ereport(ERROR, (errcode(ERRCODE_IO_ERROR),
+								errmsg("failed to COPY to shard %ld on %s:%s",
+									   shardId, nodeName, nodePort)));
+			}
+
+			continue;
+		}
+
+		/* check whether there were any COPY errors */
+		result = PQgetResult(connection);
+		if (PQresultStatus(result) != PGRES_COMMAND_OK && stopOnFailure)
+		{
+			ReportCopyError(connection, result);
+		}
+
+		PQclear(result);
+	}
+}
+
+
+/*
+ * ReportCopyError tries to report a useful error message for the user from
+ * the remote COPY error messages.
+ */
+static void
+ReportCopyError(PGconn *connection, PGresult *result)
+{
+	char *remoteMessage = PQresultErrorField(result, PG_DIAG_MESSAGE_PRIMARY);
+
+	if (remoteMessage != NULL)
+	{
+		/* probably a constraint violation, show remote message and detail */
+		char *remoteDetail = PQresultErrorField(result, PG_DIAG_MESSAGE_DETAIL);
+
+		ereport(ERROR, (errmsg("%s", remoteMessage),
+						errdetail("%s", remoteDetail)));
+	}
+	else
+	{
+		/* probably a connection problem, get the message from the connection */
+		char *lastNewlineIndex = NULL;
+
+		remoteMessage = PQerrorMessage(connection);
+		lastNewlineIndex = strrchr(remoteMessage, '\n');
+
+		/* trim trailing newline, if any */
+		if (lastNewlineIndex != NULL)
+		{
+			*lastNewlineIndex = '\0';
+		}
+
+		ereport(ERROR, (errmsg("%s", remoteMessage)));
+	}
+}
+
+
+/*
+ * ColumnOutputFunctions walks over a table's columns, and finds each column's
+ * type information. The function then resolves each type's output function,
+ * and stores and returns these output functions in an array.
+ */
+FmgrInfo *
+ColumnOutputFunctions(TupleDesc rowDescriptor, bool binaryFormat)
+{
+	uint32 columnCount = (uint32) rowDescriptor->natts;
+	FmgrInfo *columnOutputFunctions = palloc0(columnCount * sizeof(FmgrInfo));
+
+	uint32 columnIndex = 0;
+	for (columnIndex = 0; columnIndex < columnCount; columnIndex++)
+	{
+		FmgrInfo *currentOutputFunction = &columnOutputFunctions[columnIndex];
+		Form_pg_attribute currentColumn = rowDescriptor->attrs[columnIndex];
+		Oid columnTypeId = currentColumn->atttypid;
+		Oid outputFunctionId = InvalidOid;
+		bool typeVariableLength = false;
+
+		if (binaryFormat)
+		{
+			getTypeBinaryOutputInfo(columnTypeId, &outputFunctionId, &typeVariableLength);
+		}
+		else
+		{
+			getTypeOutputInfo(columnTypeId, &outputFunctionId, &typeVariableLength);
+		}
+
+		Assert(currentColumn->attisdropped == false);
```

Comment:
I suspect that this isn't ok anymore, and that we instead need to skip such columns.